File: sorting/heap.py
from util.list import *


class MaxHeap:

    def __init__(self, l):
        # time complexity
        # build a heap by inserting each element from list o(nlogn)
        # build the heap by a list o(n)
        self.list = list([None])
        # build the heap from a list 
        if l:
            for i in l:
                self.list.append(i)
            # the index of the last element that is not leaf
            # is len(l)//2 e.g. 11 -> 5
            # from the last not leaf element shift down to the first
            for i in range(len(l)//2, 0, -1):
                self.shift_down(i)

    def size(self):
        return len(self.list)

    def insert(self, item):
        self.list.append(item)
        self.shift_up()

    def pop(self):
        if len(self.list) == 1:
            return
        # get the fist element that will be removed
        ret = self.list[1]
        # remove the last element to the first position
        self.list[1] = self.list[-1]
        # remove the previous last element
        self.list.pop()
        if len(self.list) == 1:
            return ret

        self.shift_down()
        return ret

    def is_empty(self):
        return len(self.list) == 1

    def shift_up(self):
        i = len(self.list) - 1
        while i//2 >= 1 and self.list[i] > self.list[i//2]:
            swap(self.list, i, i//2)
            i = i//2
    
    def __str__(self):
        return ', '.join(str(e) for e in self.list[1:])

    def shift_down(self, index=1):
        # shift down from index
        i = index
        length = len(self.list)
        # check if i is the last elememt now
        while 2 * i <= length - 1:
            # # index of the left child
            # left_index = 2 * i
            # # index of the right child
            # right_index = left_index + 1
           
            # # left child is greater than current element
            # if self.list[left_index] > self.list[i]:

            #     if right_index < length and self.list[left_index] < self.list[right_index]:
            #         swap(self.list, right_index, i)
            #         i = right_index
            #         continue
                
            #     swap(self.list, left_index, i)
            #     i = left_index
            #     continue
            # elif right_index < length and self.list[right_index] > self.list[i]:
            #     swap(self.list, right_index, i)
            #     i = right_index
            # else:
            #     break
            k = 2 * i
            if k + 1 <= length - 1 and self.list[k + 1] > self.list[k]:
                k = k + 1
            
            if self.list[k] <= self.list[i]:
                break
            
            swap(self.list, k, i)
            i = k

if __name__ == '__main__':
    heap = MaxHeap()

    test_list = generate_random_list(100, 1, 1000)
    for e in test_list:
        heap.insert(e)

    while not heap.is_empty():
         print(heap.pop())

File: Js/heap.js
const objUtil = require('./objUtil');
const arrUtil = require('./arrUtil');

// 索引堆
let MaxHeap = function(arr) {
    this.indexArr = arrUtil.generateArrWithRange(0, arr.length - 1);
    this.dataArr = objUtil.clone(arr);
    this.size = this.dataArr.length;
    this.heapify();
}

MaxHeap.prototype.printHeap =function() {
    for(let i of this.indexArr) {
        console.log(this.dataArr[i]);
    }
}

MaxHeap.prototype.heapify = function() {
    let start = Math.floor((this.size - 2) / 2);

    for(let i = start; i >= 0; i--) {
        this.shiftDown(i);
    }
}

MaxHeap.prototype.shiftDown = function(start) {
    let i = start;
    let end = this.size - 1;

    //仍存在左节点再比较
    while(2 * i + 1 <= end) {
        let left = 2 * i + 1, right = 2 * i + 2;
        let k = left;

        if(right <= end 
           && this.dataArr[this.indexArr[right]] > this.dataArr[this.indexArr[left]]) {
               k = right;
           }

        if(this.dataArr[this.indexArr[i]] < this.dataArr[this.indexArr[k]] ) {
            // console.log(`swapping ${ i } and ${ k }`)
            arrUtil.swap(this.indexArr, i, k);
            i = k;
        }
        else break;
    } 
}

MaxHeap.prototype.shiftUp = function(i) {
    let index = i;

    while(index > 0) {
        let e = this.dataArr[this.indexArr[index]]
        let parentIndex = Math.floor((index - 1)/2);
        let parent = this.dataArr[this.indexArr[parentIndex]];

        if(parent < e) {
            arrUtil.swap(this.indexArr, index, parentIndex);
            parentIndex = index;
        }else {
            break;
        }
    }
}

MaxHeap.prototype.pop = function() {
    // 当前方法，从堆中获取元素时，只是把索引没有对原来数组进行操作
    if(this.size === 0) return;

    let ret = this.dataArr[this.indexArr[0]];

    this.indexArr[0] = this.indexArr.pop()

    this.shiftDown(0);
    this.size --;

    return ret;
}

// 把数组的第 i 个元素的值改一下
MaxHeap.prototype.change = function(i, d) {

    this.dataArr[i] = d;

    //根据索引值找到在索引数组里面对应的索引在哪
    let index;

    for(let j in this.indexArr) {
        if(this.indexArr[j] === i) {
            index = j;
            break;
        }
    }
    this.shiftUp(index);
    this.shiftDown(index);
}

MaxHeap.prototype.isEmpty = function() {
    return this.size == 0;
}

let testList = arrUtil.generateRandomList(10, 1, 10);
let maxHeap = new MaxHeap(testList);

// while(!maxHeap.isEmpty()) {
//     let e = maxHeap.pop();

//     console.log(e);
// }

maxHeap.printHeap();

maxHeap.change(1, 100);
console.log('seperate')
maxHeap.printHeap();
